internals-js/src/precompute.ts

```diff
@@ -0,0 +1,78 @@
+import {
+  assert,
+  baseType,
+  CompositeType,
+  federationMetadata,
+  FieldDefinition,
+  collectTargetFields,
+  InterfaceType,
+  ObjectType,
+  Schema,
+  DirectiveDefinition
+} from ".";
+
+export const computeShareables = (schema: Schema) => {
+  const metadata = federationMetadata(schema);
+  return computeAllFieldsWithDirective(schema, metadata?.isFed2Schema() ? metadata.shareableDirective() : undefined, true);
+};
+
+export const computeKeys = (schema: Schema) => {
+  const metadata = federationMetadata(schema);
+  return computeAllFieldsWithDirective(schema, metadata?.isFed2Schema() ? metadata.keyDirective() : undefined, false);
```

Comment:
I don't think that's correct for keys, even with the `includeOtherFieldsOnType == false` because the code of `computeAllFieldsWithDirective` still adds provided fields, which aren't keys. Also, I'm not sure why we pass `undefined` as `directive` for fed1 schemas? I'd assume a function that pre-compute all fields that are part of keys should work just fine for fed1 schema.